publish.go

```diff
@@ -0,0 +1,314 @@
+	log.Debugf("Uploading file %v with package name '%v', AppEditId '%v", appFile, packageName, appEditID)
+	editsApksService := androidpublisher.NewEditsApksService(service)
+
+	editsApksUploadCall := editsApksService.Upload(packageName, appEditID)
+	editsApksUploadCall.Media(appFile, googleapi.ContentType("application/octet-stream"))
+
+	apk, err := editsApksUploadCall.Do()
+	if err != nil {
+		return &androidpublisher.Apk{}, fmt.Errorf("failed to upload apk, error: %s", err)
+	}
+	log.Infof("Uploaded apk version: %d", apk.VersionCode)
+	return apk, nil
+}
+
+// updates the listing info of a given release.
+func updateListing(configs Configs, release *androidpublisher.TrackRelease) error {
+	log.Debugf("Checking if updating listing is required, whats new dir is '%v'", configs.WhatsnewsDir)
+	if configs.WhatsnewsDir != "" {
+		fmt.Println()
+		log.Infof("Update listing started")
+
+		recentChangesMap, err := readLocalisedRecentChanges(configs.WhatsnewsDir)
+		if err != nil {
+			return fmt.Errorf("failed to read whatsnews, error: %s", err)
+		}
+
+		var releaseNotes []*androidpublisher.LocalizedText
+		for language, recentChanges := range recentChangesMap {
+			releaseNotes = append(releaseNotes, &androidpublisher.LocalizedText{
+				Language:        language,
+				Text:            recentChanges,
+				ForceSendFields: []string{},
+				NullFields:      []string{},
+			})
+		}
+		release.ReleaseNotes = releaseNotes
+		log.Infof("Update listing finished")
+	}
+	return nil
+}
+
+// readLocalisedRecentChanges reads the recent changes from the given path and returns them as a map.
+func readLocalisedRecentChanges(recentChangesDir string) (map[string]string, error) {
+	recentChangesMap := map[string]string{}
+
+	pattern := filepath.Join(recentChangesDir, "whatsnew-*-*")
+	recentChangesPaths, err := filepath.Glob(pattern)
+	if err != nil {
+		return map[string]string{}, err
+	}
+
+	pattern = `whatsnew-(?P<local>.*-.*)`
+	re := regexp.MustCompile(pattern)
+
+	for _, recentChangesPath := range recentChangesPaths {
+		matches := re.FindStringSubmatch(recentChangesPath)
+		if len(matches) == 2 {
+			language := matches[1]
+			content, err := fileutil.ReadStringFromFile(recentChangesPath)
+			if err != nil {
+				return map[string]string{}, err
+			}
+
+			recentChangesMap[language] = content
+		}
+	}
+	if len(recentChangesMap) > 0 {
+		log.Debugf("Found the following recent changes:")
+		for language, recentChanges := range recentChangesMap {
+			log.Debugf("%v\n", language)
+			log.Debugf("Content: %v", recentChanges)
+		}
+	} else {
+		log.Debugf("No recent changes found")
+	}
+
+	return recentChangesMap, nil
+}
+
+// GetRelease gets a release from a track based on it's status.
+func GetRelease(config Configs, releases *[]*androidpublisher.TrackRelease) *androidpublisher.TrackRelease {
+	status := GetReleaseStatusFromConfig(config)
+	for _, release := range *releases {
+		if status == release.Status {
```

Comment:
Can we be sure that there is only one release with a given status? IN case if there can be multiple ones, can we get the latest one? In the first case leave a comment explaining this.

Reply:
Added comment.